prow/spyglass/lenses/metadata/lens.go

```diff
@@ -137,6 +146,79 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
 	return buf.String()
 }
 
+var failedMountRegex = regexp.MustCompile(`MountVolume.SetUp failed for volume "(.+?)" : (.+?)`)
+
+func hintFromPodInfo(buf []byte) string {
+	var report k8sreporter.PodReport
+	if err := json.Unmarshal(buf, &report); err != nil {
+		// This error isn't worth highlighting here, and will come up again
+		// elsewhere.
+		return ""
+	}
+
+	// We're more likely to pick a relevant event if we use the last ones first.
+	sort.Slice(report.Events, func(i, j int) bool {
+		a := &report.Events[i]
+		b := &report.Events[j]
+		return b.LastTimestamp.Before(&a.LastTimestamp)
+	})
+
+	// If the pod completed successfully there's probably not much to say.
+	if report.Pod.Status.Phase == v1.PodSucceeded {
+		return ""
+	}
+	// Check if we have any images that didn't pull
+	for _, s := range append(report.Pod.Status.InitContainerStatuses, report.Pod.Status.ContainerStatuses...) {
+		if s.State.Waiting != nil && s.State.Waiting.Reason == "ImagePullBackOff" {
+			return fmt.Sprintf("The %s container could not start because it could not pull %q. Check your images.", s.Name, s.Image)
+		}
+	}
+	// Check if we're trying to mount a volume
+	if report.Pod.Status.Phase == v1.PodPending {
+		failedMount := false
+		for _, e := range report.Events {
+			if e.Reason == "FailedMount" {
+				failedMount = true
+				if strings.HasPrefix(e.Message, "MountVolume.SetUp") {
+					// Annoyingly, parsing this message is the only way to get this information.
+					// If we can't parse it, we'll fall through to a generic bad volume message below.
+					results := failedMountRegex.FindStringSubmatch(e.Message)
+					if results == nil {
+						continue
+					}
+					return fmt.Sprintf("The pod could not start because it could not mount the volume %q: %s", results[1], results[2])
+				}
+			}
+		}
+		if failedMount {
+			return fmt.Sprintf("The job could not start because one or more of the volumes could not be mounted.")
+		}
+	}
+	// Check if we cannot be scheduled
+	// This is unlikely - we only outright fail if a pod is actually scheduled to a pod it does not support
+	if report.Pod.Status.Phase == v1.PodFailed && report.Pod.Status.Reason == "MatchNodeSelector" {
+		return fmt.Sprintf("The job could not start because it was scheduled to a node that does not satisfy its NodeSelector")
+	}
+	// Usually we would fail to schedule it at all, so it will be pending forever.
+	if report.Pod.Status.Phase == v1.PodPending {
+		for _, e := range report.Events {
+			if e.Reason == "FailedScheduling" {
+				return fmt.Sprintf("There are no nodes that your pod can schedule to - check your requests, tolerations, and node selectors (%s)", e.Message)
+			}
+		}
+	}
+
+	// There are a bunch of fun ways for the node to fail that we've seen before
```

Comment:
We probably should also gate this with a check that the pod is `pending` or `failed`. Some of these might be transient errors that pods can recover from. The same applies to the ImagePullBackoff check above.
Alternatively, above where we abort if the pod phase is `success`, we could also abort if the pod phase is `running` so long as none of the errors we want to alert on present themselves on `running` pods.

Reply:
We only report podinfo.json for completed jobs. If your job has completed but your pod is still running, something has probably gone wrong, at which point it may be worth reporting on these failures even if the pod did start (but not end) successfully.

I don't feel terribly strongly about this one way or the other, though.

re ImagePullBackoff — in that case we're checking the _current_ state of the containers (at time of job termination), not the historical events, so it is definitionally impossible for the pod to be running while the containers are still in ImagePullBackoff.